### ISSUE TITLE
prediction: define a canonical multimodal runtime forecast contract

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#8049]
Parent issue: #8047
Blocked by: none
Related consumers: #8051, #8062, #5307, #2844

```yaml
archetype: implementation
evidence_tier: idea
linked_policy:
  - docs/context/issue_1512_issue_archetypes.md
  - docs/context/artifact_evidence_vocabulary.md
```

## Problem

The current runtime prediction contract in `robot_sf/nav/predictive_types.py` represents one Gaussian trajectory distribution per pedestrian:

```text
ProbabilisticPrediction
  by_pedestrian[id] -> TrajectoryDistribution(mean, cov, confidence)
```

That is sufficient for the constant-velocity baseline, but not for a planner that must distinguish mutually exclusive pedestrian hypotheses such as continue, inferred desired motion, and yield/decelerate. #5307 and #2844 also need a stable way to exchange multiple modes without each defining a private schema.

This issue must establish the **single canonical runtime representation** for multimodal pedestrian forecasts while preserving all current unimodal callers.

## Scope and ownership

Primary owner:

- `robot_sf/nav/predictive_types.py`

Inspect before editing:

- `robot_sf/nav/baseline_probabilistic_predictor.py`
- the canonical forecast batch/schema used by #2844 and benchmark forecast tooling
- `robot_sf/planner/chance_constrained_mpc.py`
- `robot_sf/planner/risk_aware_trajectory_ranker.py`
- tests constructing `ProbabilisticPrediction` directly

Do **not** introduce an unrelated planner-private forecast class or a third serialization format. Runtime types may gain a multimodal container, but benchmark serialization must adapt to/from the canonical owner rather than fork it.

## Required runtime design

Implement an immutable, validated representation equivalent to:

```python
@dataclass(frozen=True)
class TrajectoryMode:
    mode_id: str
    probability: float
    trajectory: TrajectoryDistribution
    provenance: str

@dataclass(frozen=True)
class PedestrianForecast:
    track_id: int
    modes: tuple[TrajectoryMode, ...]
    existence_probability: float = 1.0
    source_confidence: float = 1.0
    age_steps: int = 0

@dataclass(frozen=True)
class MultimodalPrediction:
    dt: float
    horizon: int
    by_pedestrian: Mapping[int, PedestrianForecast]
```

Exact names may change only if an existing canonical owner already defines better names. Required semantics:

1. `mode_id` is stable, non-empty, and unique within one pedestrian forecast; it is not an array index.
2. Mode probabilities are finite, non-negative, and sum to one within a documented tolerance.
3. Every mode has exactly the container horizon and timestep.
4. Mean/covariance retain `TrajectoryDistribution` shape, finite, and PSD validation.
5. `existence_probability` and `source_confidence` are finite in `[0, 1]`.
6. `age_steps` is a non-negative integer.
7. Track identity is explicit and must agree with the dictionary key; mismatch is an error.
8. An empty global prediction is valid; a pedestrian forecast with zero modes is invalid.
9. Track and mode ordering for serialization/diagnostics is canonical and deterministic, never inherited from observation-row order.
10. Validated forecasts cannot be mutated through source-array aliases; use defensive copies/read-only arrays according to repository convention.

## Backward compatibility

Keep `TrajectoryDistribution` and `ProbabilisticPrediction` source-compatible for current callers.

Add one explicit adapter such as:

```python
def as_multimodal_prediction(
    prediction: ProbabilisticPrediction | MultimodalPrediction,
    *,
    default_mode_id: str = "unimodal",
    provenance: str = "legacy_adapter",
) -> MultimodalPrediction:
    ...
```

A legacy distribution becomes exactly one mode with probability `1.0`; `dt`, horizon, mean, covariance, and confidence remain numerically unchanged. Accepted input types/failures must be explicit; do not use broad duck typing.

## Probability normalization policy

The core dataclass must reject arbitrary malformed probabilities rather than silently repairing them.

If predictors need a convenience builder, provide a separately named helper that:

- requires finite non-negative weights;
- requires a positive finite total;
- records the pre-normalization total;
- normalizes deterministically;
- applies only a documented small tolerance/floor;
- rejects all-zero and non-finite input.

The serialized forecast must contain final normalized probabilities, not hidden unnormalized logits.

## Serialization and evidence contract

Provide one JSON-safe export method/helper containing at least:

- schema/version identifier;
- `dt` and horizon;
- ordered track IDs;
- per-track existence, source confidence, and age;
- ordered mode IDs, probabilities, and provenance;
- mean and covariance arrays.

Before freezing field names, locate the forecast artifact owner used by #2844. If a schema already exists, extend it compatibly or add an explicit versioned adapter. Record the decision in the PR. A silent incompatible artifact change is out of scope.

Reject NaN/Inf rather than relying on permissive JSON output. Use stable key/array ordering and repository canonical hash/serialization helpers where available.

## Failure behavior

Raise specific `ValueError`/`TypeError` failures for:

- empty or duplicate mode IDs;
- negative, NaN, or infinite probabilities;
- probability sum outside tolerance;
- mode/container horizon or `dt` mismatch;
- invalid trajectory mean/covariance;
- key/track-ID mismatch;
- invalid existence/confidence/age;
- unsupported legacy input type;
- mutation/alias assumptions that violate immutability.

Do not drop one bad mode and continue with a forecast that appears complete.

## Tests

Add focused tests, likely in `tests/test_predictive_types.py` or the adjacent canonical owner:

1. Valid one-track/two-mode construction.
2. Multi-track canonical ordering independent of input dictionary order.
3. Legacy prediction becomes one probability-1 mode without numerical drift.
4. Duplicate/empty mode ID rejected.
5. Negative/NaN/Inf probability rejected.
6. Probability sum below/above tolerance rejected.
7. Builder normalization is deterministic and rejects all-zero weights.
8. Mismatched horizon rejected.
9. Mismatched `dt` rejected where represented at mode level.
10. Malformed/non-PSD covariance rejected through the existing owner.
11. Invalid existence/confidence/age rejected.
12. Track-key mismatch rejected.
13. Source-array mutation after construction cannot mutate the forecast.
14. JSON-safe export is deterministic and rejects non-finite values.
15. Empty global prediction remains valid.
16. Existing unimodal tests pass unchanged.
17. Current baseline probabilistic predictor runs through the adapter end to end.
18. An exported/loaded fixture preserves mode IDs/probabilities/arrays under the selected schema version.

## Definition of done

- [ ] One canonical multimodal runtime representation exists in the prediction owner.
- [ ] Existing unimodal callers remain source-compatible.
- [ ] A named unimodal-to-multimodal adapter preserves values exactly.
- [ ] Track/mode ordering is deterministic and documented.
- [ ] Validation covers shapes, finiteness, covariance, probabilities, identity, horizon, and metadata.
- [ ] Serialization is versioned or explicitly adapted to the existing forecast artifact owner.
- [ ] #2844 and #5307 can consume the contract without private mode containers.
- [ ] No planner behavior/default or pinned release artifact changes.
- [ ] Unit tests and Ruff pass.

## Validation / Testing

```bash
ruff check robot_sf/nav/predictive_types.py robot_sf/nav/baseline_probabilistic_predictor.py tests
pytest -q tests/test_predictive_types.py
pytest -q <existing baseline probabilistic predictor tests>
```

## Non-goals

- Implementing force-residual prediction; #8051 owns it.
- Choosing collision-risk aggregation; #8062 owns it.
- Changing planner selection.
- Training/evaluating #2844.
- Migrating pinned release artifacts.

## Stop conditions

Update this issue before continuing if:

- the existing forecast artifact schema cannot be adapted without a breaking version change;
- #5307 has a conflicting canonical multimodal type not visible from current `main`;
- preserving `ProbabilisticPrediction` compatibility proves impossible;
- multimodality would require simulator-hidden information.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 8049
source_body_sha256: 07e6e7f79de0f5e57ba8296ecfc285a952979e6c707aad22ed187d24ee046c56
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: state_conflict
next_action: reconcile_state_labels
authority: lifecycle_owner
execution_mode: lifecycle
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-05
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
